sdks/python/apache_beam/runners/interactive/pipeline_instrument.py

```diff
@@ -393,6 +398,45 @@ def _replace_with_cached_inputs(self, pipeline):
     cache, noop.
     """
 
+    # Find all cached unbounded PCollections.
+    class CacheableUnboundedPCollectionVisitor(PipelineVisitor):
+      def __init__(self, pin):
+        self._pin = pin
+        self.unbounded_pcolls = set()
+
+      def enter_composite_transform(self, transform_node):
+        self.visit_transform(transform_node)
+
+      def visit_transform(self, transform_node):
+        if transform_node.inputs:
+          for input_pcoll in transform_node.inputs:
+            key = self._pin.cache_key(input_pcoll)
+            if (key in self._pin._cached_pcoll_read and
+                not input_pcoll.is_bounded):
```

Comment:
Similar to the above comment, this `is_bounded` might work for some unbounded sources (extending `io.base`), but it cannot cover all scenarios.

Reply:
It does. The is_bounded is even checked in the core.py for GroupAlsoByWindow. The FlinkStreamingImpulseSource is such an example that doesn't extend io.base but still creates an unbounded PCollection.

Reply:
I see. Thanks, Sam! I'm going to mark the comments resolved. As long as the source implementation sets the `is_bounded` field correctly either through the `io.base` `is_bounded` or directly invoking `pvalue.PCollection(pbegin.pipeline, is_bounded=False)
 `, we can count on it to check the unboundedness.

I was worried about the default value being `is_bounded=True` and all unbounded source implementation need to set the value to `False` correctly "everywhere" without any enforcement. And there is no mechanism to truly check if a PCollection is bounded or not.

Reply:
Great, thank you for being careful and thorough!